### PR TITLE
api: Make getServerSettings robust to server bug with missing realm_name

### PR DESCRIPTION
**Pull request by chrisbobbe**
Discussion: https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/.2Fserver_settings.3A.20.60realm_name.60.2C.20etc.2E/near/1334042

### CODE REVIEW

File: src/api/settings/getServerSettings.js [hunk 1]
```diff
@@ -44,5 +61,21 @@ export type ApiResponseServerSettings = {|
 |};
 
 /** See https://zulip.com/api/get-server-settings */
-export default async (realm: URL): Promise<ApiResponseServerSettings> =>
-  apiGet({ apiKey: '', email: '', realm }, 'server_settings');
+export default async (realm: URL): Promise<ApiResponseServerSettings> => {
+  const result = await apiGet({ apiKey: '', email: '', realm }, 'server_settings');
+  if (result.realm_name == null) {
```
Comment (gnprice) [inline src/api/settings/getServerSettings.js:66]: This means that the object this function actually returns always does have a `realm_name`, right?

Let's express that in the return type. That will let this function more fully act as a "crunchy shell", avoiding the need for the changes in other files.

We can separately express the type of what we actually expect from the server, following the pattern of other functions like `api.getMessages`, like this:
```js
type ServerApiResponseServerSettings = {|
  ...ApiResponseServerSettings,

  // … comment …
  realm_name?: string,
|};
```

Then say `const result: ServerApiResponseServerSettings = await apiGet(…);`, and Flow can check the types from there.

**Review by gnprice — COMMENTED** [review]
Thanks @chrisbobbe! One comment on the implementation strategy below.

Have you tested this empirically? I'm curious what the UX ends up being.

**Comment by chrisbobbe** [comment]
Thanks for the review! Revision pushed.

> Have you tested this empirically? I'm curious what the UX ends up being.

I've just tested at the new revision. In `getServerSettings`, in place of `await apiGet(…)`, I dropped in [this example response](https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/.2Fserver_settings.3A.20.60realm_name.60.2C.20etc.2E/near/1332909) that Anders gave with the comment

> With the default `ROOT_DOMAIN_LANDING_PAGE = False` and the root realm deleted, I get this from https://andersk.zulipdev.org/api/v1/server_settings:

Then I entered "chat.zulip.org" in the realm-input screen and got this:



Then I tried picking a logged-out account from the account picker and got this:

**Comment by chrisbobbe** [comment]
> I dropped in [this example response](https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/.2Fserver_settings.3A.20.60realm_name.60.2C.20etc.2E/near/1332909)

Note that more than `realm_name` is missing here; `realm_icon` and `realm_description` are also missing. That's as expected, I guess; the [server code comment](https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/.2Fserver_settings.3A.20.60realm_name.60.2C.20etc.2E/near/1332888) says "etc.":

> realm_name, realm_icon, etc. are not guaranteed to appear in the response.

but I think it's probably fine to just look at `realm_name`.

File: src/api/settings/getServerSettings.js [hunk 2]
```diff
@@ -52,6 +53,47 @@ export type ApiResponseServerSettings = {|
...
-  apiGet({ apiKey: '', email: '', realm }, 'server_settings');
+export default async (realm: URL): Promise<ApiResponseServerSettings> => {
+  const result: ServerApiResponseServerSettings = await apiGet(
+    { apiKey: '', email: '', realm },
+    'server_settings',
+  );
+
+  const { realm_name } = result;
+  if (realm_name == null) {
+    // See comment on realm_name.
```
Comment (gnprice) [inline src/api/settings/getServerSettings.js:85]: nit: this sounds like it means the local variable of that name; "on realm_name in ServerApiResponseServerSettings" would make it clear.

**Review by gnprice — COMMENTED** [review]
Thanks for the revision and that testing!

Merging, after fixing one nit below.

**Pull request merged**